### ISSUE TITLE
Support for MPI jobs (JobSet extensions)

### ISSUE THREAD

**Issue by danielvegamyhre** [kubernetes-sigs/jobset#146]
**UPDATE**: see https://github.com/kubernetes-sigs/jobset/issues/146#issuecomment-1912813909 for latest update to the goals of this issue

**What would you like to be added**:
Official, tested support for MPI jobs. Some things we need to decide:
- Will the user or JobSet controller be responsible for creating the ssh config (Secrets)?
- Will the user or JobSet controller be responsible for setting up the environment variables configuring slots per host, etc?

These questions can also be addressed more broadly in the design for configuration defaulting for certain workload types (PyTorch job, TensorFlow job, MPI job, etc.)

**Why is this needed**:
MPI is a popular parallel computing paradigm with implementations such as OpenMPI, IntelMPI that is commonly used for HPC jobs, and can be modeled as JobSet (this was prototyped and confirmed during the JobSet research and prototyping phase).

**Comment by kannon92** [comment]
Hey @vsoch. Have you got a working example with MPI and JobSet?  I know you are doing some benchmarks for K8s/HPC so was just curious if you got MPI working with jobSet.

**Comment by vsoch** [comment]
> Hey @vsoch. Have you got a working example with MPI and JobSet? I know you are doing some benchmarks for K8s/HPC so was just curious if you got MPI working with jobSet.

I have several! I have a prototype of the Flux Operator with JobSet (but it's not easy to inspect) and just this weekend I put together OSU benchmarks and (a private app) called Netmark, both which are MPI based. [Here](https://converged-computing.github.io/metrics-operator/getting_started/metrics.html) is the little description / registry page, and tutorials to each (for Python and command line):

The objects are created by the operator, so they aren't explicitly there. Do you want me to throw one together (preferences?) and I can send you all the YAML for it? 

- https://github.com/converged-computing/metrics-operator/tree/main/examples/python/network-netmark
- https://github.com/converged-computing/metrics-operator/tree/main/examples/python/network-osu-benchmark

**Comment by kannon92** [comment]
I'd love an example with YAML.

**Comment by vsoch** [comment]
okey doke! Here is a quick one! This is an older variant of what the operator makes but it should work. Note that:

- the headless service is needed for the network
- the hostnames are derived (ip addresses) for the job that is given in a text file list to mpirun
- the communication happens via ssh
- config maps provide the entrypoints for worker/launcher
- you will probably want to add resource limits/requests so there is one worker or launcher per node
- you'll also want to adjust the tasks based on your resources
- OSU benchmarks are point to point and intended to be run on two nodes (one launcher, one worker) this is different than many/most MPI apps. My operator checks that the number of pods is exactly 2 but there is no limit here.

Let me know if you have other question, happy to help! Also send along any metrics you like to run for kubernetes (IO/storage/performance/network etc.) I'm adding them to this operator for easy deployment!
[osu-benchmarks.yaml.txt](https://github.com/kubernetes-sigs/jobset/files/12284531/osu-benchmarks.yaml.txt)

**Comment by vsoch** [comment]
And here is a complete updated config with the logging updates I made to the metrics operator:

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  creationTimestamp: "2023-08-07T23:15:22Z"
  generation: 1
  name: metricset-sample
  namespace: default
  ownerReferences:
  - apiVersion: flux-framework.org/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: MetricSet
    name: metricset-sample
    uid: 6e33f7f4-aa3a-415a-94de-da352cb17e55
  resourceVersion: "778"
  uid: 319f7408-d5ac-4c67-b590-7b3d8a15d938
spec:
  failurePolicy: {}
  network:
    enableDNSHostnames: false
    subdomain: ms
  replicatedJobs:
  - name: m
    replicas: 1
    template:
      metadata:
        name: metricset-sample
        namespace: default
      spec:
        activeDeadlineSeconds: 31500000
        backoffLimit: 100
        completionMode: NonIndexed
        completions: 1
        parallelism: 1
        template:
          metadata:
            labels:
              app.kubernetes.io/name: metricset-sample
              cluster-name: metricset-sample
              metricset-name: metricset-sample
              namespace: default
            name: metricset-sample
            namespace: default
          spec:
            containers:
            - command:
              - /bin/bash
              - /metrics_operator/entrypoint-0.sh
              image: ghcr.io/converged-computing/metric-sysstat:latest
              imagePullPolicy: IfNotPresent
              name: io-sysstat
              resources: {}
              stdin: true
              tty: true
              volumeMounts:
              - mountPath: /metrics_operator/
                name: metricset-sample
                readOnly: true
              - mountPath: /workflow
                name: storage
            restartPolicy: OnFailure
            setHostnameAsFQDN: true
            shareProcessNamespace: false
            subdomain: ms
            volumes:
            - configMap:
                items:
                - key: entrypoint-0
                  mode: 511
                  path: entrypoint-0.sh
                name: metricset-sample
              name: metricset-sample
            - hostPath:
                path: /tmp/workflow
              name: storage
  successPolicy:
    operator: All
  suspend: false
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: metricset-sample
  namespace: default
data:
  entrypoint-0: "#!/bin/bash\ni=0\necho \"METADATA START {\\\"pods\\\":1,\\\"completions\\\":1,\\\"storageVolumePath\\\":\\\"/workflow\\\",\\\"storageVolumeHostPath\\\":\\\"/tmp/workflow\\\",\\\"metricName\\\":\\\"io-sysstat\\\",\\\"metricDescription\\\":\\\"statistics
    for Linux tasks (processes) : I/O, CPU, memory, etc.\\\",\\\"metricType\\\":\\\"storage\\\",\\\"metricOptions\\\":{\\\"completions\\\":2,\\\"human\\\":\\\"false\\\",\\\"rate\\\":10}}\nMETADATA
    END\"\ncompletions=2\necho \"METRICS OPERATOR COLLECTION START\"\nwhile true\n
    \ do\n    echo \"METRICS OPERATOR TIMEPOINT\"\n\tiostat -dxm -o JSON\n\t# Note
    we can do iostat -o JSON\n\tif [[ $completions -ne 0 ]] && [[ $i -eq $completions
    ]]; then\n    \techo \"METRICS OPERATOR COLLECTION END\"\n    \texit 0\n    fi\n\tsleep
    10\n\tlet i=i+1\ndone\n"
---
apiVersion: v1
kind: Service
metadata:
  name: ms
  namespace: default
spec:
  clusterIP: None
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  selector:
    metricset-name: metricset-sample
  sessionAffinity: None
  type: ClusterIP
```

TLDR it makes it easier to parse the log into a meaningful output, e.g., just parsed and made these plots:

https://github.com/converged-computing/operator-experiments/tree/main/google/service-timing/run14/c2-node-60-240#analyze

**Comment by kannon92** [comment]
I was able to play around with your example and it worked!  I do  think some thoughts will be needed on how to make this more general for different number of ranks.

**Comment by vsoch** [comment]
Awesome! Actually that's what the metrics operator does - that particular dump of YAML above is generated from:

```yaml
apiVersion: flux-framework.org/v1alpha1
kind: MetricSet
metadata:
  labels:
    app.kubernetes.io/name: metricset
    app.kubernetes.io/instance: metricset-sample
  name: metricset-sample
spec:
  # OSU benchmark is point to point and MUST be run with 2 pods
  pods: 2
  metrics:
   - name: network-osu-benchmark
```
But also, the OSU benchmarks do not run on more than two ranks. They will throw up on you an error message if you try. If you are interested in that, we have a tool called netmark that makes nice matrices for point to point latency across an entire cluster and are planning to make it public (hopefully) in the next few months!  https://github.com/converged-computing/google-batch/blob/main/netmark/img/netmark-experiment-size-100-heatmap.pdf

![image](https://github.com/kubernetes-sigs/jobset/assets/814322/96134580-946e-4feb-97a7-22faf6f45a17)

**Comment by vsoch** [comment]
@danielvegamyhre to give some high level comments - for this to work from JobSet, you'd need to add the same logic that we do for the metrics-operator above, or the MPI operator. At the simplest level, you need to have a main pod in the JobSet acting as a launcher (the one running the command) and then the rest are workers. They need to be configured to accept ssh. I don't think ssh is the best way to bootstrap (e.g., flux uses zeromq which we've shown to be faster) but it's a first shot. The design I'd advocate for here isn't something hard coded into JobSet, but rather an easy ability to define some pattern that can map a request to run a command to this design, and install the needed ssh stuff that is missing. 

We definitely have models of this working with JobSet - several in fact, and I'm interested to hear your thoughts on such a general approach that would be provided natively (if I understand the gist of this issue correctly).

**Comment by danielvegamyhre** [comment]
> @danielvegamyhre to give some high level comments - for this to work from JobSet, you'd need to add the same logic that we do for the metrics-operator above, or the MPI operator. At the simplest level, you need to have a main pod in the JobSet acting as a launcher (the one running the command) and then the rest are workers. They need to be configured to accept ssh. I don't think ssh is the best way to bootstrap (e.g., flux uses zeromq which we've shown to be faster) but it's a first shot. The design I'd advocate for here isn't something hard coded into JobSet, but rather an easy ability to define some pattern that can map a request to run a command to this design, and install the needed ssh stuff that is missing.
> 
> We definitely have models of this working with JobSet - several in fact, and I'm interested to hear your thoughts on such a general approach that would be provided natively (if I understand the gist of this issue correctly).

Thanks for your thoughts on this! Yeah I prototyped MPI workloads using k8s primitives last year when we were originally researching the requirements for various ML and HPC workload types, to gather the requirements for JobSet and validate the API design.

To prototype MPI using k8s primitives I used an indexed Job, headless service, Secret for ssh key, and a configMaps for hostfile values. However, coming up with a design that can support all this and keeps JobSet generic/flexible and not too bespoke, is pretty challenging. So we prioritized support for other features that were being more commonly requested by users.

Now we are planning to work on support for these more complex workload configurations this year. I discussed this with @ahg-g recently and one idea he had was to support "extensions," where the user can specify an extension in their JobSet spec, and the webhooks + controllers will have points where they call a hook for whatever extensions the JobSet has, and the hook will perform mutation logic (env var configurations, k8s object creation, or whatever else) that needs to be done for this workload type. This way we can avoid intermingling the core JobSet controller logic with logic that is specific to particular workload types.

**Comment by vsoch** [comment]
That's awesome! That last paragraph is literally the design of the metrics-operator - the idea was to use JobSet like legos, adding on whatever additional volumes / commands/ features are needed. And it worked pretty well, although I probably went a bit too deep in terms of wanting to play with an interface design. Please ping me if there is something fun to collaborate on.

**Comment by vsoch** [comment]
Actually I just realized @ahg-g was literally sitting on the stage for my talk, so he saw (some of the high level) description about the need for easy to deploy common patterns for jobs. Really excited that you are going to work on this, and please include me if you are able.
